bootstrap/src/main/java/io/airlift/bootstrap/LifeCycleManager.java

```diff
@@ -221,14 +238,49 @@ public void addInstance(Object instance)
         }
     }
 
+    private void stopInstance(Object obj, LifeCycleStopFailureHandler handler)
+    {
+        log.debug("Stopping %s", obj.getClass().getName());
+        LifeCycleMethods methods = methodsMap.get(obj.getClass());
+        for (Method preDestroy : methods.methodsFor(PreDestroy.class)) {
+            log.debug("\t%s()", preDestroy.getName());
+            try {
+                preDestroy.invoke(obj);
+            }
+            catch (Exception e) {
+                handler.handlePreDestroyException(obj.getClass(), preDestroy, unwrapInvocationTargetException(e));
+            }
+        }
+    }
+
     private void startInstance(Object obj)
-            throws IllegalAccessException, InvocationTargetException
+            throws Exception
     {
         log.debug("Starting %s", obj.getClass().getName());
         LifeCycleMethods methods = methodsMap.get(obj.getClass());
         for (Method postConstruct : methods.methodsFor(PostConstruct.class)) {
             log.debug("\t%s()", postConstruct.getName());
-            postConstruct.invoke(obj);
+            try {
+                postConstruct.invoke(obj);
+            }
+            catch (Exception e) {
+                Exception cause = unwrapInvocationTargetException(e);
+                log.debug(cause, "Exception in PostConstruct method %s::%s()", obj.getClass().getName(), postConstruct.getName());
```

Comment:
Rather than logging here and suffering from log-and-throw, what if we enhance the thrown exception:
```java
LifecycleStartException failure = new LifecycleStartException(
        format("Exception in PostConstruct method %s::%s()", obj.getClass().getName(), postConstruct.getName()),
        unwrapInvocationTargetException(e));
stopInstance(obj, (klass, method, exception) -> {
    String message = format("Exception in PreDestroy method %1$s::%2$s() after PostConstruct failure in %1$s::%3$s()", klass.getName(), method.getName(), postConstruct.getName());
    failure.addSuppressed(new RuntimeException(message, exception));
});
throw failure;
```
With this change, we'd introduce a new unchecked `LifecycleStartException` and change the `start()` method to throw that for all failures. This aligns it with `stop()` for the same reasons -- callers don't have any reasonable way to handle the underlying exception.